src/app/clusters/ota-requestor/OTARequestor.cpp

```diff
@@ -377,36 +365,52 @@ void OTARequestor::mOnConnected(void * context, chip::DeviceProxy * deviceProxy)
         if (deviceProxy != nullptr)
         {
             chip::Messaging::ExchangeManager * exchangeMgr = deviceProxy->GetExchangeManager();
-            chip::Optional<chip::SessionHandle> session    = deviceProxy->GetSecureSession();
-            if (exchangeMgr != nullptr && session.HasValue())
+            if (exchangeMgr == nullptr)
             {
-                exchangeCtx = exchangeMgr->NewContext(session.Value(), &mBdxMessenger);
+                ChipLogError(SoftwareUpdate, "Cannot retrieve exchange manager from device");
+                return;
             }
 
-            if (exchangeCtx == nullptr)
+            chip::Optional<chip::SessionHandle> session = deviceProxy->GetSecureSession();
+            if (!session.HasValue())
             {
-                ChipLogError(BDX, "unable to allocate ec: exchangeMgr=%p sessionExists? %u", exchangeMgr, session.HasValue());
+                ChipLogError(SoftwareUpdate, "Cannot retrieve session from device");
+                return;
+            }
+
+            requestorCore->mExchangeCtx = exchangeMgr->NewContext(session.Value(), &requestorCore->mBdxMessenger);
+
+            if (requestorCore->mExchangeCtx == nullptr)
+            {
+                ChipLogError(BDX, "Unable to allocate ec: exchangeMgr=%p sessionExists? %u", exchangeMgr, session.HasValue());
                 return;
             }
         }
 
-        mBdxMessenger.Init(mBdxDownloader, exchangeCtx);
-        mBdxDownloader->SetMessageDelegate(&mBdxMessenger);
-        CHIP_ERROR err = mBdxDownloader->SetBDXParams(initOptions);
+        requestorCore->mBdxMessenger.Init(requestorCore->mBdxDownloader, requestorCore->mExchangeCtx);
+        requestorCore->mBdxDownloader->SetMessageDelegate(&requestorCore->mBdxMessenger);
+        CHIP_ERROR err = requestorCore->mBdxDownloader->SetBDXParams(initOptions);
         VerifyOrReturn(err == CHIP_NO_ERROR,
                        ChipLogError(SoftwareUpdate, "Error init BDXDownloader: %" CHIP_ERROR_FORMAT, err.Format()));
-        err = mBdxDownloader->BeginPrepareDownload();
+        err = requestorCore->mBdxDownloader->BeginPrepareDownload();
         VerifyOrReturn(err == CHIP_NO_ERROR,
-                       ChipLogError(SoftwareUpdate, "Error init BDXDownloader: %" CHIP_ERROR_FORMAT, err.Format()));
+                       ChipLogError(SoftwareUpdate, "Cannot begin prepare download: %" CHIP_ERROR_FORMAT, err.Format()));
         break;
     }
     default:
         break;
     }
 }
 
+// Called whenever FindOrEstablishSession fails
+void OTARequestor::OnConnectionFailure(void * context, NodeId deviceId, CHIP_ERROR error)
+{
+    ChipLogError(SoftwareUpdate, "Failed to connect to node 0x%" PRIX64 ": %" CHIP_ERROR_FORMAT, deviceId, error.Format());
+}
+
 void OTARequestor::TriggerImmediateQuery()
 {
-    // Perhaps we don't need a separate function ConnectToProvider, revisit this
-    ConnectToProvider();
+    ConnectToProvider(kQueryImage);
```

Comment:
If this is called from an unsafe session context, this will cause concurrency problems. ConnectToProvider must be called from a known safe context, but is expected to be called by the Application. At the very least, method docs should clearly indicate that this must be called from the same externally synchronized context as any other Matter stack method

Reply:
Added a comment in the header file for this method